src/v3/plugins/github/graphView.js

```diff
@@ -0,0 +1,298 @@
+    try {
+      this._checkInvariants();
+    } finally {
+      this._isCheckingInvariants = false;
+    }
+  }
+
+  _checkInvariants() {
+    // Check the node type <-> prefix mapping is maintained
+    const nodeTypeToInvariant = {
+      [GN.REPO_TYPE]: {hasParent: false, parentAccessor: null},
+      [GN.ISSUE_TYPE]: {hasParent: true, parentAccessor: (x) => x.repo},
+      [GN.PULL_TYPE]: {hasParent: true, parentAccessor: (x) => x.repo},
+      [GN.COMMENT_TYPE]: {hasParent: true, parentAccessor: (x) => x.parent},
+      [GN.REVIEW_TYPE]: {hasParent: true, parentAccessor: (x) => x.pull},
+      [GN.USERLIKE_TYPE]: {hasParent: false, parentAccessor: null},
+    };
+    for (const node of this._graph.nodes({prefix: GN._Prefix.base})) {
+      const structuredNode = GN.fromRaw((node: any));
+      const type = structuredNode.type;
+      const invariant = nodeTypeToInvariant[type];
+      if (invariant == null) {
+        throw new Error(`Invariant: Unexpected node type ${type}`);
+      }
+      const {hasParent, parentAccessor} = invariant;
+      if (hasParent) {
+        // Will throw invariant error if there is not exactly 1 parent
+        const parent = this.parent((structuredNode: any));
+        if (parentAccessor == null) {
+          throw new Error(`Impossible; make flow happy`);
+        }
+        const expectedParent = parentAccessor((structuredNode: any));
+        if (!deepEqual(parent, expectedParent)) {
+          throw new Error(`${stringify(structuredNode)} has the wrong parent`);
+        }
+      }
+    }
+
+    type Hom = {|
+      +srcPrefix: NodeAddressT,
+      +dstPrefix: NodeAddressT,
+    |};
+    function homProduct(
+      srcPrefixes: GN.RawAddress[],
+      dstPrefixes: GN.RawAddress[]
+    ): Hom[] {
+      const result = [];
+      for (const srcPrefix of srcPrefixes) {
+        for (const dstPrefix of dstPrefixes) {
+          result.push({srcPrefix, dstPrefix});
+        }
+      }
+      return result;
+    }
+    type EdgeInvariant = {|
+      +homs: Hom[],
+      +srcAccessor?: (GE.StructuredAddress) => NodeAddressT,
+      +dstAccessor?: (GE.StructuredAddress) => NodeAddressT,
+    |};
+    const edgeTypeToInvariants: {[type: string]: EdgeInvariant} = {
+      [GE.HAS_PARENT_TYPE]: {
+        homs: [
+          {srcPrefix: GN._Prefix.issue, dstPrefix: GN._Prefix.repo},
+          {srcPrefix: GN._Prefix.pull, dstPrefix: GN._Prefix.repo},
+          {srcPrefix: GN._Prefix.review, dstPrefix: GN._Prefix.pull},
+          {srcPrefix: GN._Prefix.reviewComment, dstPrefix: GN._Prefix.review},
+          {srcPrefix: GN._Prefix.issueComment, dstPrefix: GN._Prefix.issue},
+          {srcPrefix: GN._Prefix.pullComment, dstPrefix: GN._Prefix.pull},
+        ],
+        srcAccessor: (x) => GN.toRaw((x: any).child),
+      },
+      [GE.MERGED_AS_TYPE]: {
+        homs: [
+          {
+            srcPrefix: GN._Prefix.pull,
+            dstPrefix: NodeAddress.fromParts(["sourcecred", "git", "COMMIT"]),
```

Comment:
Interestingly, this API exists, but is not public. What should we do
here? Should we make `_Prefix` public?